### PR TITLE
MM-45755 - Calls: Unmute automatically in DM or GM channels

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- This makes mobile behave the same as webapp.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-45755

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
- Android: 12, Pixel 6
- iOS: 15.3.1, iPhone 7 plus

#### Release Note
```release-note
Calls: unmute automatically in DMs and GMs
```

### CODE REVIEW

File: app/products/calls/components/leave_and_join_alert.tsx [hunk 1]
```diff
@@ -75,5 +76,14 @@ export const doJoinCall = async (serverUrl: string, channelId: string, intl: Int
...
         errorAlert(res.error?.toString() || seeLogs, intl);
+        return;
+    }
+
+    if (isDMorGM) {
+        // FIXME (MM-46048) - HACK
+        // There's a race condition between unmuting and receiving existing tracks from other participants.
+        // Fixing this properly requires extensive and potentially breaking changes.
+        // Waiting for a second before unmuting is a decent workaround that should work in most cases.
+        setTimeout(() => unmuteMyself(), 1000);
```
Comment (enahum) [inline app/products/calls/components/leave_and_join_alert.tsx:87]: Add this in an useEffect and then clean the timeout on unmount?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/products/calls/components/leave_and_join_alert.tsx (hunk 1 above)
Comment (cpoile) [reply app/products/calls/components/leave_and_join_alert.tsx]: @enahum Normally yes, but in this case there isn't a react component to useEffect in -- it's a straight function. (But even if there were, you would still have joined the call, and you would still want the `unmute` to fire.) Does that make sense?

I think having this function in the components folder is probably a mistake and leads to confusion. I will refactor it. Thanks!

**Review by enahum — APPROVED** [review]
Cool it now makes sense

**Review by streamer45 — APPROVED** [review]

**Comment by streamer45** [comment]
@cpoile Interestingly starting a new call in a DM on my android test device is not triggering this behaviour. I was able to trace the problem to the `unmute()` call returning early due to `peer` being `null`. I am thinking there may be some timing issue there, maybe we are not waiting for everything to be ready before returning from `joinCall`?

**Comment by cpoile** [comment]
@streamer45 That's weird, it definitely works on my Pixel. The code shouldn't reach the unmute call if the connection hasn't completed, so peer should not be null... can you try building and installing from scratch?

**Comment by streamer45** [comment]
> @streamer45 That's weird, it definitely works on my Pixel. The code shouldn't reach the unmute call if the connection hasn't completed, so peer should not be null... can you try building and installing from scratch?

That's what I did but I'll try again. Joining an existing call works fine so it's definitely weird.

**Comment by cpoile** [comment]
@streamer45 Should be fixed, can you check again? Thanks

**Pull request merged**